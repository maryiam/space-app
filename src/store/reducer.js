import * as a from './actions';
import uniqBy from 'lodash/uniqBy';
import map from 'lodash/map';
import { combineReducers } from 'redux';

const initialPayment = {
  list: [],
};

export function payment(state = initialPayment, action) {
  switch (action.type) {
    case a.FIND_PAYMENT.REQUEST:
      return {
        ...state,
        loading: true,
      };
    case a.FIND_PAYMENT.SUCCESS:
      return {
        ...state,
        loading: false,
        list: uniqBy([...state.list, ...action.list], 'id'),
        found: map(action.list, 'id'),
      };
    case a.FIND_PAYMENT.FAILURE:
      return {
        ...state,
        loading: false,
        found: [],
      };
    default:
      return state;
  }
}

export function order(state = {}, action) {
  switch (action.type) {
    case a.RESET_DESTINATION:
      return {
        ...state,
        destination: null,
      };
    case a.SET_DESTINATION:
      return {
        ...state,
        destination: action.destinationId,
      };
    case a.SET_PAYMENT:
      return {
        ...state,
        payment: action.paymentId,
      };
    case a.SUBMIT:
      return {
        ...state,
        departing: true,
      };
    default:
      return state;
  }
}

export function destination(state = { list: [] }, action) {
  switch (action.type) {
    case a.SET_SEARCH:
      return {
        ...state,
        search: action.searchCriteria,
        loading: true,
      };
    case a.FIND_DESTINATION.SUCCESS:
      return {
        ...state,
        loading: false,
        list: uniqBy(action.list, 'id'),
      };
    case a.FIND_DESTINATION.FAILURE:
      return {
        ...state,
        loading: false,
        list: [],
      };
    default:
      return state;
  }
}

export default combineReducers({
  destination,
  payment,
  order,
});
