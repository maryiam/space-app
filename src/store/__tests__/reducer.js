import { destination, payment, order } from '../reducer';

describe('store/reducer', () => {
  describe('#payment', () => {
    it('reduces a payment/FIND/REQUEST action', () => {
      const action = {
        type: 'payment/FIND/REQUEST',
        currencies: [0, 1],
      };
      const state = payment(undefined, action);
      expect(state).toMatchSnapshot();
    });

    it('reduces a payment/FIND/SUCCESS action', () => {
      const action = {
        type: 'payment/FIND/SUCCESS',
        list: [{ id: 0, name: 'test 1' }, { id: 1, name: 'test 2' }],
      };
      const state = payment(undefined, action);
      expect(state).toMatchSnapshot();
    });

    it('reduces a payment/FIND/FAILURE action', () => {
      const action = {
        type: 'payment/FIND/FAILURE',
        error: 'failure',
      };
      const state = payment(undefined, action);
      expect(state).toMatchSnapshot();
    });

    it('reduces an unknown action', () => {
      const action = {
        type: 'UNKNOWN',
      };
      const state = payment(undefined, action);
      expect(state).toMatchSnapshot();
    });
  });

  describe('#order', () => {
    it('reduces a order/RESET_DESTINATION action', () => {
      const action = {
        type: 'order/RESET_DESTINATION',
      };
      const state = order(undefined, action);
      expect(state).toMatchSnapshot();
    });

    it('reduces a order/SET_DESTINATION action', () => {
      const action = {
        type: 'order/SET_DESTINATION',
        destinationId: 0,
      };
      const state = order(undefined, action);
      expect(state).toMatchSnapshot();
    });

    it('reduces a order/SET_PAYMENT action', () => {
      const action = {
        type: 'order/SET_PAYMENT',
        paymentId: 0,
      };
      const state = order(undefined, action);
      expect(state).toMatchSnapshot();
    });

    it('reduces a order/SUBMIT action', () => {
      const action = {
        type: 'order/SUBMIT',
      };
      const state = order(undefined, action);
      expect(state).toMatchSnapshot();
    });

    it('reduces an unknown action', () => {
      const action = {
        type: 'UNKNOWN',
      };
      const state = order(undefined, action);
      expect(state).toMatchSnapshot();
    });
  });

  describe('#destination', () => {
    it('reduces a destination/SET_SEARCH action', () => {
      const action = {
        type: 'destination/SET_SEARCH',
        searchCriteria: 'Co',
      };
      const state = destination(undefined, action);
      expect(state).toMatchSnapshot();
    });

    it('reduces a destination/FIND/SUCCESS action', () => {
      const action = {
        type: 'destination/FIND/SUCCESS',
        list: [
          { id: 0, name: 'test 1', currencies: [0, 1] },
          { id: 1, name: 'test 2', currencies: [2] },
        ],
      };
      const state = destination(undefined, action);
      expect(state).toMatchSnapshot();
    });

    it('reduces a destination/FIND/FAILURE action', () => {
      const action = {
        type: 'destination/FIND/FAILURE',
        error: 'failed',
      };
      const state = destination(undefined, action);
      expect(state).toMatchSnapshot();
    });

    it('reduces an unknown action', () => {
      const action = {
        type: 'UNKNOWN',
      };
      const state = destination(undefined, action);
      expect(state).toMatchSnapshot();
    });
  });
});
