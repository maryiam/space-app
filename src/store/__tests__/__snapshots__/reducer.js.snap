// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store/reducer #destination reduces a destination/FIND/FAILURE action 1`] = `
Object {
  "list": Array [],
  "loading": false,
}
`;

exports[`store/reducer #destination reduces a destination/FIND/SUCCESS action 1`] = `
Object {
  "list": Array [
    Object {
      "currencies": Array [
        0,
        1,
      ],
      "id": 0,
      "name": "test 1",
    },
    Object {
      "currencies": Array [
        2,
      ],
      "id": 1,
      "name": "test 2",
    },
  ],
  "loading": false,
}
`;

exports[`store/reducer #destination reduces a destination/SET_SEARCH action 1`] = `
Object {
  "list": Array [],
  "loading": true,
  "search": "Co",
}
`;

exports[`store/reducer #destination reduces an unknown action 1`] = `
Object {
  "list": Array [],
}
`;

exports[`store/reducer #order reduces a order/RESET_DESTINATION action 1`] = `
Object {
  "destination": null,
}
`;

exports[`store/reducer #order reduces a order/SET_DESTINATION action 1`] = `
Object {
  "destination": 0,
}
`;

exports[`store/reducer #order reduces a order/SET_PAYMENT action 1`] = `
Object {
  "payment": 0,
}
`;

exports[`store/reducer #order reduces a order/SUBMIT action 1`] = `
Object {
  "departing": true,
}
`;

exports[`store/reducer #order reduces an unknown action 1`] = `Object {}`;

exports[`store/reducer #payment reduces a payment/FIND/FAILURE action 1`] = `
Object {
  "found": Array [],
  "list": Array [],
  "loading": false,
}
`;

exports[`store/reducer #payment reduces a payment/FIND/REQUEST action 1`] = `
Object {
  "list": Array [],
  "loading": true,
}
`;

exports[`store/reducer #payment reduces a payment/FIND/SUCCESS action 1`] = `
Object {
  "found": Array [
    0,
    1,
  ],
  "list": Array [
    Object {
      "id": 0,
      "name": "test 1",
    },
    Object {
      "id": 1,
      "name": "test 2",
    },
  ],
  "loading": false,
}
`;

exports[`store/reducer #payment reduces an unknown action 1`] = `
Object {
  "list": Array [],
}
`;
